File: packages/go/hushspec/sink.go
package hushspec

import (
	"encoding/json"
	"fmt"
	"os"
)

// ReceiptSink persists or forwards decision receipts.
type ReceiptSink interface {
	Send(receipt *DecisionReceipt) error
}

// PolicyEventSink is a sink that can also record which policy is in force
// (log spec 6).
//
// It is a separate, optional interface rather than a second method on
// [ReceiptSink] so that every sink written against the 0.1 API keeps
// compiling; [RecordPolicyEvent] routes an event to a sink that implements it
// and drops it for one that does not.
type PolicyEventSink interface {
	ReceiptSink
	RecordPolicyEvent(event *PolicyEvent) error
}

// RecordPolicyEvent writes a policy-in-effect record to sink when it can carry
// one, and reports whether it did. A plain [ReceiptSink] -- stderr, a filter,
// a callback over receipts -- has nowhere to put the event and is not an
// error: only a hash-linked log needs the record to tie receipts to the policy
// that produced them.
func RecordPolicyEvent(sink ReceiptSink, event *PolicyEvent) (bool, error) {
	target, ok := sink.(PolicyEventSink)
	if !ok {
		return false, nil
	}
	return true, target.RecordPolicyEvent(event)
}

// FileReceiptSink appends receipts as JSON Lines to a file.
type FileReceiptSink struct {
	path string
}

func NewFileReceiptSink(path string) *FileReceiptSink {
	return &FileReceiptSink{path: path}
}

func (s *FileReceiptSink) Send(receipt *DecisionReceipt) error {
	f, err := os.OpenFile(s.path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return fmt.Errorf("sink: open file: %w", err)
	}
	defer f.Close()

	data, err := json.Marshal(receipt)
	if err != nil {
		return fmt.Errorf("sink: marshal receipt: %w", err)
	}

	_, err = fmt.Fprintf(f, "%s\n", data)
	if err != nil {
		return fmt.Errorf("sink: write receipt: %w", err)
	}

	return nil
}

// StderrReceiptSink writes pretty-printed receipts to stderr.
type StderrReceiptSink struct{}

func (s *StderrReceiptSink) Send(receipt *DecisionReceipt) error {
	data, err := json.MarshalIndent(receipt, "", "  ")
	if err != nil {
		return fmt.Errorf("sink: marshal receipt: %w", err)
	}

	fmt.Fprintf(os.Stderr, "[hushspec] %s\n", data)
	return nil
}

// FilteredSink only forwards receipts matching the configured decisions.
type FilteredSink struct {
	inner     ReceiptSink
	decisions []Decision
}

func NewFilteredSink(sink ReceiptSink, decisions []Decision) *FilteredSink {
	return &FilteredSink{
		inner:     sink,
		decisions: decisions,
	}
}

func NewDenyOnlySink(sink ReceiptSink) *FilteredSink {
	return NewFilteredSink(sink, []Decision{DecisionDeny})
}

func (s *FilteredSink) Send(receipt *DecisionReceipt) error {
	for _, d := range s.decisions {
		if receipt.Decision == d {
			return s.inner.Send(receipt)
		}
	}
	return nil
}

// RecordPolicyEvent forwards the event whatever the filter is: the filter
// selects which *decisions* are worth keeping, and a policy event is what ties
// the kept receipts to the policy that produced them.
func (s *FilteredSink) RecordPolicyEvent(event *PolicyEvent) error {
	_, err := RecordPolicyEvent(s.inner, event)
	return err
}

// MultiSink fans out to all sinks. Returns the first error but always
// attempts every sink.
type MultiSink struct {
	sinks []ReceiptSink
}

func NewMultiSink(sinks []ReceiptSink) *MultiSink {
	return &MultiSink{sinks: sinks}
}

func (s *MultiSink) Send(receipt *DecisionReceipt) error {
	var firstErr error
	for _, sink := range s.sinks {
		if err := sink.Send(receipt); err != nil {
			if firstErr == nil {
				firstErr = err
			}
		}
	}
	return firstErr
}

// RecordPolicyEvent fans the event out to every sink that can carry one,
// returning the first error but always attempting each sink.
func (s *MultiSink) RecordPolicyEvent(event *PolicyEvent) error {
	var firstErr error
	for _, sink := range s.sinks {
		if _, err := RecordPolicyEvent(sink, event); err != nil {
			if firstErr == nil {
				firstErr = err
			}
		}
	}
	return firstErr
}

// CallbackSink invokes a function for each receipt.
type CallbackSink struct {
	callback func(*DecisionReceipt) error
}

func NewCallbackSink(callback func(*DecisionReceipt) error) *CallbackSink {
	return &CallbackSink{callback: callback}
}

func (s *CallbackSink) Send(receipt *DecisionReceipt) error {
	return s.callback(receipt)
}

// NullSink discards all receipts.
type NullSink struct{}

func (s *NullSink) Send(receipt *DecisionReceipt) error {
	return nil
}

// RecordPolicyEvent discards the event, like everything else a NullSink is
// handed.
func (s *NullSink) RecordPolicyEvent(event *PolicyEvent) error {
	return nil
}

File: crates/hushspec/src/sink.rs
use crate::evaluate::Decision;
use crate::receipt::DecisionReceipt;
use std::fs::OpenOptions;
use std::io::Write;
use std::path::{Path, PathBuf};

#[derive(Debug, thiserror::Error)]
pub enum SinkError {
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),
    #[error("serialization error: {0}")]
    Serialization(#[from] serde_json::Error),
    /// A hash-linked log could not be continued or written consistently.
    #[error("log chain error: {0}")]
    Chain(String),
}

/// Where an engine puts the receipts it records.
///
/// A sink must never break enforcement: implementations report a failure as
/// [`SinkError`] rather than panicking, and callers surface it on the
/// `sink.error` observer channel instead of denying on it.
pub trait ReceiptSink: Send + Sync {
    /// Record one decision receipt.
    ///
    /// # Errors
    ///
    /// Whatever the destination reports; see [`SinkError`].
    fn send(&self, receipt: &DecisionReceipt) -> Result<(), SinkError>;

    /// Record a policy-in-effect event (log spec 6). Sinks that only carry
    /// receipts ignore it; the hash-linked log writes it as an entry.
    fn record_policy_event(&self, _event: &crate::log::PolicyEvent) -> Result<(), SinkError> {
        Ok(())
    }

    /// How a `sink.error` observer event names this sink in its `source`.
    ///
    /// Defaults to the implementing type's own name; override it with
    /// something an operator can place, such as a path or an endpoint.
    fn name(&self) -> &'static str {
        short_type_name(std::any::type_name::<Self>())
    }
}

/// The last segment of a fully qualified type path, with any generic
/// arguments left off: `hushspec::sink::FilteredSink` is `FilteredSink`.
fn short_type_name(path: &'static str) -> &'static str {
    let bare = path.split('<').next().unwrap_or(path);
    bare.rsplit("::").next().unwrap_or(bare)
}

/// Appends receipts as JSON Lines to a file.
pub struct FileReceiptSink {
    path: PathBuf,
}

impl FileReceiptSink {
    /// Append receipts to `path`, creating it on the first write.
    #[must_use]
    pub fn new(path: impl AsRef<Path>) -> Self {
        Self {
            path: path.as_ref().to_path_buf(),
        }
    }
}

impl ReceiptSink for FileReceiptSink {
    fn send(&self, receipt: &DecisionReceipt) -> Result<(), SinkError> {
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)?;
        let mut record = serde_json::to_string(receipt)?;
        record.push('\n');
        file.write_all(record.as_bytes())?;
        Ok(())
    }
}

/// Pretty-prints receipts to stderr, prefixed with `[hushspec]`.
pub struct StderrReceiptSink;

impl ReceiptSink for StderrReceiptSink {
    fn send(&self, receipt: &DecisionReceipt) -> Result<(), SinkError> {
        let json = serde_json::to_string_pretty(receipt)?;
        eprintln!("[hushspec] {}", json);
        Ok(())
    }
}

/// Only forwards receipts matching configured decisions.
pub struct FilteredSink {
    inner: Box<dyn ReceiptSink>,
    decisions: Vec<Decision>,
}

impl FilteredSink {
    /// Forward only the receipts whose decision is in `decisions`.
    #[must_use]
    pub fn new(sink: Box<dyn ReceiptSink>, decisions: Vec<Decision>) -> Self {
        Self {
            inner: sink,
            decisions,
        }
    }

    /// Forward only denials.
    #[must_use]
    pub fn deny_only(sink: Box<dyn ReceiptSink>) -> Self {
        Self::new(sink, vec![Decision::Deny])
    }
}

impl ReceiptSink for FilteredSink {
    fn send(&self, receipt: &DecisionReceipt) -> Result<(), SinkError> {
        if self.decisions.contains(&receipt.decision) {
            self.inner.send(receipt)
        } else {
            Ok(())
        }
    }
}

/// Fans out to multiple sinks. Returns the first error but invokes all sinks.
pub struct MultiSink {
    sinks: Vec<Box<dyn ReceiptSink>>,
}

impl MultiSink {
    /// Fan every receipt out to all of `sinks`.
    #[must_use]
    pub fn new(sinks: Vec<Box<dyn ReceiptSink>>) -> Self {
        Self { sinks }
    }
}

impl ReceiptSink for MultiSink {
    fn send(&self, receipt: &DecisionReceipt) -> Result<(), SinkError> {
        let mut first_error: Option<SinkError> = None;
        for sink in &self.sinks {
            if let Err(e) = sink.send(receipt)
                && first_error.is_none()
            {
                first_error = Some(e);
            }
        }
        first_error.map_or(Ok(()), Err)
    }

    fn record_policy_event(&self, event: &crate::log::PolicyEvent) -> Result<(), SinkError> {
        let mut first_error: Option<SinkError> = None;
        for sink in &self.sinks {
            if let Err(e) = sink.record_policy_event(event)
                && first_error.is_none()
            {
                first_error = Some(e);
            }
        }
        first_error.map_or(Ok(()), Err)
    }
}

/// No-op sink. `send()` always succeeds.
pub struct NullSink;

impl ReceiptSink for NullSink {
    fn send(&self, _receipt: &DecisionReceipt) -> Result<(), SinkError> {
        Ok(())
    }
}

type SinkCallback = dyn Fn(&DecisionReceipt) -> Result<(), SinkError> + Send + Sync;

/// Invokes a closure for each receipt.
pub struct CallbackSink {
    callback: Box<SinkCallback>,
}

impl CallbackSink {
    /// Hand each receipt to `callback`.
    #[must_use]
    pub fn new(
        callback: impl Fn(&DecisionReceipt) -> Result<(), SinkError> + Send + Sync + 'static,
    ) -> Self {
        Self {
            callback: Box::new(callback),
        }
    }
}

impl ReceiptSink for CallbackSink {
    fn send(&self, receipt: &DecisionReceipt) -> Result<(), SinkError> {
        (self.callback)(receipt)
    }
}
